pkg/kubelet/dockershim/exec.go

```diff
@@ -106,7 +108,7 @@ func (*NativeExecHandler) ExecInContainer(ctx context.Context, client libdocker.
 		ExecStarted:  execStarted,
 	}
 
-	if timeout > 0 {
+	if timeout > 0 && utilfeature.DefaultFeatureGate.Enabled(features.ExecProbeTimeout) {
```

Comment:
The thinking here was that the timeout is always respected, but the kubelet just ignores it from the prober:

https://github.com/kubernetes/kubernetes/blob/4bb30c3b0e9167eb85b1e3b0c9a3bab3b351a3e2/pkg/probe/exec/exec.go#L72-L79

Is that behavior not happening?

Reply:
Also, again: we've only observed this `ExecProbeTimeout=false` pathology in dockershim-enabled clusters (hence the change in this surface are). So this is a docker-only thing as far as we've discovered.

Reply:
> What's happening in the ExecProbeTimeout=false configuration is that (1) the timeout being reached does not initiate a container reset and (2) if the actual exec command fails after the configured timeout configuration then similarly the container reset does not happen.

Sorry I think I'm still missing context -- I _think_ this the old behavior we're trying to preserve? Prior to implementing exec probe timeouts, this was the behavior from dockershim:
* ignore probe timeouts entirely, where timedout liveness probe do not restart containers
* if the exec probe fails after the timeout (or after 10 seconds), it is not evaluated from the prober

Reply:
Agree 100% the intent is to preserve the pre-1.20.0 dockershim behavior.

What we see doesn't match the 2nd bullet point above. If the exec probe fails after the timeout, it *is* evaluated from the prober.

I have a PR in the aks-engine project that shows this happening:

https://github.com/Azure/aks-engine/pull/4325

Basically, the new tests pass in 1.18 and 1.19, but fail in 1.20. The one that's important is this one:

https://github.com/Azure/aks-engine/pull/4325/files#diff-0bd2bb48395375d835916863ebd4b2bf05747e18d16d8e551588bec8a447b470

The above spec checks out in 1.18 and 1.19: pod restarts are observed on 5+ second failures despite the fact that the 1 second timeout is "ignored".

Reply:
> What we see doesn't match the 2nd bullet point above. If the exec probe fails after the timeout, it is evaluated from the prober.

Ah, this is the part I was missing :) So a probe failing after the timeout does trigger a container restart when ExecProbeTimeout=false?

Reply:
We've tested 1.18 and 1.19 (there is no ExecProbeTimeout feature flag there). The behavior in 1.18 and 1.19 is that a failing probe that takes longer than a timeout triggers a container restart. That's what we want 1.20+ and `ExecProbeTimeout=false` to replicate.

But what we see is that 1.20+ and `ExecProbeTimeout=false` do not restart containers when the probe fails after the timeout period.

tl;dr for dockershim: (pre-1.20) compared to (1.20 + `ExecProbeTimeout=false`) are not equivalent.

Reply:
Gotcha, thanks for clarifying.

> The behavior in 1.18 and 1.19 is that a failing probe that takes longer than a timeout triggers a container restart.

One clarifying question on this, do you know if this is true for timeouts greater than 10 seconds?

Reply:
I'll test and report back!

Reply:
10+ second probe failure restarts the container as expected in 1.19.8:

```
STEP: Validating that the exec livenessProbe caused at least one pod restart due to command failure, even if timeout is not respected
2021/03/29 13:15:43 $ k describe pod exec-liveness-timeout-always-fail -n default
2021/03/29 13:15:44 #### $ k describe pod exec-liveness-timeout-always-fail -n default completed in 256.907866ms
2021/03/29 13:15:44 
Name:         exec-liveness-timeout-always-fail
Namespace:    default
Priority:     0
Node:         k8s-agentpool1-28118473-vmss000001/10.240.0.65
Start Time:   Mon, 29 Mar 2021 13:14:56 -0700
Labels:       test=liveness
Annotations:  kubernetes.io/psp: privileged
Status:       Running
IP:           10.240.0.68
IPs:
  IP:  10.240.0.68
Containers:
  exec-liveness-timeout-always-fail:
    Container ID:  docker://6a9a11af16d487e3722cfa33e2565550f425f70d5c726da6d69b3e9bea337b5b
    Image:         busybox:1.31.1
    Image ID:      docker-pullable://busybox@sha256:95cf004f559831017cdf4628aaf1bb30133677be8702a8c5f2994629f637a209
    Port:          <none>
    Host Port:     <none>
    Args:
      /bin/sh
      -c
      while true; do rm -f /tmp/healthy; sleep 5; done
    State:          Running
      Started:      Mon, 29 Mar 2021 13:15:43 -0700
    Last State:     Terminated
      Reason:       Error
      Exit Code:    137
      Started:      Mon, 29 Mar 2021 13:15:28 -0700
      Finished:     Mon, 29 Mar 2021 13:15:43 -0700
    Ready:          True
    Restart Count:  3
    Liveness:       exec [sh -c sleep 11 && test -f /tmp/healthy] delay=1s timeout=1s period=1s #success=1 #failure=1
    Environment:    <none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-fzsh5 (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  default-token-fzsh5:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-fzsh5
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  kubernetes.io/os=linux
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason     Age               From               Message
  ----     ------     ----              ----               -------
  Normal   Scheduled  47s               default-scheduler  Successfully assigned default/exec-liveness-timeout-always-fail to k8s-agentpool1-28118473-vmss000001
  Warning  Unhealthy  3s (x3 over 33s)  kubelet            Liveness probe failed:
  Normal   Killing    3s (x3 over 33s)  kubelet            Container exec-liveness-timeout-always-fail failed liveness probe, will be restarted
  Normal   Pulled     1s (x4 over 46s)  kubelet            Container image "busybox:1.31.1" already present on machine
  Normal   Created    1s (x4 over 46s)  kubelet            Created container exec-liveness-timeout-always-fail
  Normal   Started    0s (x4 over 46s)  kubelet            Started container exec-liveness-timeout-always-fail
```

Reply:
The current change looks good to me for preserving pre-1.20 behavior. Rather than checking for TimeoutError now to warn about exceeding the timeout, should we now check the actual time it took to execute the probe and warn if that exceeded the configured timeout?

As a follow-up, in 1.22 (and maybe picked back to 1.21 / 1.20), I'd actually like to see a metric published for probes whose execution time exceeds their timeout (or maybe exceeds 90% of the timeout or something) so admins can determine whether it is safe to set ExecProbeTimeout=true.